### DISCUSSION TITLE
Mu should support Circup for package management in CircuitPython Mode

### DISCUSSION THREAD

**Discussion by ntoll** [mu-editor/mu#1332] [Ideas & New Features]
[Circup](https://github.com/adafruit/circup) is a `pip` like tool for managing packages on CircuitPython devices. @makermelissa suggested that Mu expose some of Circup's core features since it's a friendly thing to do.

Things `circup` can do (from the command help):

```
$ circup
Usage: circup [OPTIONS] COMMAND [ARGS]...

  A tool to manage and update libraries on a CircuitPython device.

Options:
  --verbose         Comprehensive logging is sent to stdout.
  --version         Show the version and exit.
  --path DIRECTORY  Path to CircuitPython directory. Overrides automatic
                    path detection.
  --help            Show this message and exit.
  -r --requirement  Supports requirements.txt tracking of library
                    requirements with freeze and install commands.

Commands:
  freeze        Output details of all the modules found on the connected...
  install       Install a named module(s) onto the device.
  list          Lists all out of date modules found on the connected...
  show          Show the long list of all available modules in the bundle.
  show <query>  Search the names in the modules in the bundle for a match.
  uninstall     Uninstall a named module(s) from the connected device.
  update        Update modules on the device. Use --all to automatically update
                all modules.
```

At the very least I think Mu should be able to:

* List modules on the connected device (and mark those which are out of date, along with an update feature). Users should also have an "uninstall" feature too.
* Show all modules currently available in the library bundle and allow users to select those they want to install.
* Find modules given a search string, displaying their current status on the device and/or availability in the bundle (with options to install/update/uninstall).

Something for the admin pane when in CircuitPython mode..?

cc/ @tannewt @kattni @DaveSprague @carlosperate @fmorton

**Comment by carlosperate** [comment]
Sounds great!

The ESP mode has a flasher in the settings which could be a good starting point for how this feature might look like:
<img width="720" alt="image" src="https://user-images.githubusercontent.com/4189262/109130771-016ae500-774a-11eb-8a84-3c747cd1231d.png">


> Find modules given a search string, displaying their current status on the device and/or availability in the bundle (with options to install/update/uninstall).

With the aim to simplify things, is this feature really needed?
If the user can see what they have installed (and updated the packages/modules) and there is also a dropdown with all the modules they can install, what can they do with this search that they cannot do already?

**Reply by dybber** [reply]
There's also work done towards a package installer for MicroPython, as MicroPython uses pip to manage packages it's very similar to the "Third Party Packages"-tab in Python-mode. See the Work-in-progress here: https://github.com/mu-editor/mu/issues/1053

**Comment by carlosperate** [comment]
How does circup find the modules btw? Is there an API endpoint? Or does it download a zip file with all of them?
If it downloads a zip file somewhere, is this something we want to bundle with Mu? (Maybe not as it will get staled quickly?)

**Reply by ntoll** [reply]
It downloads details from a known location.

**Reply by carlosperate** [reply]
But is it details about the packages (just metadata to do queries), or does it downaload a zipped bundle with all the packages in it (and then reads the data from those)?

**Reply by tannewt** [reply]
It downloads a zip from here: https://github.com/adafruit/Adafruit_CircuitPython_Bundle/releases/tag/20210227n
